pkg/querier/remote/read.go

```diff
@@ -0,0 +1,157 @@
+package remote
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/config"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logcli/client"
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/querier"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/stores/index/stats"
+)
+
+// ReadConfig is the configuration for reading from remote storage.
+// code from /github.com/prometheus/prometheus/config/config.go:868
+type ReadConfig struct {
+	URL   *config.URL `yaml:"url"`
+	Name  string      `yaml:"name,omitempty"`
+	OrgID string      `yaml:"orgID,omitempty"`
+
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	//todo:@liguozhong support http herder
+	Headers map[string]string `yaml:"headers,omitempty"`
+	//todo:@liguozhong support RemoteTimeout
+	RemoteTimeout time.Duration `yaml:"remote_timeout,omitempty"`
+}
+
+func NewQuerier(name string, remoteReadConfig ReadConfig) (querier.Querier, error) {
+	client := &client.DefaultClient{
+		OrgID:   remoteReadConfig.OrgID,
+		Address: remoteReadConfig.URL.String(),
+		TLSConfig: config.TLSConfig{
+			InsecureSkipVerify: remoteReadConfig.HTTPClientConfig.TLSConfig.InsecureSkipVerify,
+		},
+	}
+	if remoteReadConfig.HTTPClientConfig.BasicAuth != nil {
+		client.Username = remoteReadConfig.HTTPClientConfig.BasicAuth.Username
+		client.Password = strings.TrimSpace(string(remoteReadConfig.HTTPClientConfig.BasicAuth.Password))
+	}
+
+	return &Querier{
+		client: client,
+		name:   name,
+	}, nil
+}
+
+type Querier struct {
+	client *client.DefaultClient
+	codec  queryrangebase.Codec
+	name   string
+}
+
+func (q Querier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+	response, err := q.client.QueryRange(params.Selector, int(params.Limit), params.Start, params.End, params.Direction, 0, 0, false)
+	if err != nil {
+		return nil, err
+	}
+	if response.Status != loghttp.QueryStatusSuccess {
+		return nil, errors.Errorf("remote read Querier selectLogs fail,response.Status %v", response.Status)
+	}
+
+	streams, ok := response.Data.Result.(loghttp.Streams)
+	if !ok {
+		return nil, errors.New("remote read Querier selectLogs fail,value cast (loghttp.Streams) fail")
+	}
+	return iter.NewStreamsIterator(streams.ToProto(), params.Direction), nil
```

Comment:
I think we should inject the remote id like we did with the tenant ID for [multi tenant queries](https://github.com/grafana/loki/blob/8d9f162460f307d25fb4305eeccdcb0f24fbcbbb/pkg/querier/multi_tenant_querier.go#L294).

Reply:
agree , you are right

Reply:
done.

Reply:
Sorry. I cannot find it :see_no_evil: